driver/ipfs/write_batch.go

```
@@ -0,0 +1,56 @@
+package ipfs
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type WriteBatch struct {
+	db     *DB
+	wbatch *leveldb.Batch
+}
+
+func (w *WriteBatch) Put(key, value []byte) {
+	//fmt.Printf("ipfs.WriteBatch.Put key=%s , v=%s\n", string(key), string(value))
+	buf := bytes.NewBuffer(value)
+	cid, err := w.db.remoteShell.Add(buf)
+	if err != nil {
```

If writing to ipfs fails here and cid is empty, you should not continue with the subsequent code flow, and you need to report the error upward.